doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,697 @@
+  counts from such an effort should also help with multi-node performance
+  when you are able to run on a big system.
+
+
+Overall Flow
+~~~~~~~~~~~~
+
+The general flow of performance optimization will be:
+
+ * use techniques like timing the different phases of your computation to
+   try to understand which parts are the slow or not getting faster as you
+   add more locales or cores
+ * come up with a theory as to what could be improved by thinking about
+   what is happening when your program is running and/or by looking at
+   timing or communication count information
+ * make an adjustment (the adjustment only needs to be good enough to
+   measure its impact on performance at this step)
+ * measure its impact on performance
+ * if it is good, clean up the adjustment, perform correctness testing,
+   and commit/PR it to your program's revision control; if not, try
+   something else
+ * repeat
+
+Configuration Matters
+---------------------
+
+The first thing to check is that you are using Chapel in its
+highest-performing configuration for your system.
+
+ * use ``--fast`` when compiling Chapel programs for performance to tell
+   the compiler to optimize instead of adding correctness checks
+
+ * use the default configuration rather than the quickstart configuration
+
+   .. note::
+
+     You can show the current configuration with ``printchplenv`` or
+     ``chpl --print-chpl-settings``
+
+   * In particular, ``CHPL_TASKS=qthreads`` is generally faster than
+     ``CHPL_TASKS=fifo``, and ``CHPL_TARGET_MEM=jemalloc`` is generally
+     faster than ``CHPL_TARGET_MEM=cstdlib`` (see also
+     :ref:`readme-chplenv.CHPL_TASKS` and
+     :ref:`readme-chplenv.CHPL_TARGET_MEM`)
+
+   * ``CHPL_TARGET_COMPILER=llvm`` might or might not be faster than
+     using the C backend with something like
+     ``CHPL_TARGET_COMPILER=gnu``; however the performance of the LLVM
+     backend is more reliable (see also
+     :ref:`readme-chplenv.CHPL_COMPILER`)
+
+     * if you are using ``CHPL_LLVM=system``, it's a good idea to match the
+       version of LLVM bundled in the Chapel release if possible as this
+       has recieved the most attention and testing (see also
+       :ref:`readme-chplenv.CHPL_LLVM`)
+
+ * For multi-locale programs, use a high-performance networking configuration
+
+   * The following configurations support portability and correctness
+     testing but **are NOT designed for high performance**:
+
+     * ``CHPL_COMM=gasnet`` ``CHPL_COMM_SUBSTRATE=udp``
+
+     * ``CHPL_COMM=gasnet`` ``CHPL_COMM_SUBSTRATE=mpi``
+
+   * High-performance networking configuration will depend on your
+     system:
+
+     * For HPE Cray EX, use ``CHPL_COMM=ofi`` (see also :ref:`readme-cray`)
+     * For InfiniBand systems, use ``CHPL_COMM=gasnet``
+       ``CHPL_COMM_SUBSTRATE=ibv`` (see also :ref:`readme-infiniband`)
+     * For Omni-Path systems, use ``CHPL_COMM=gasnet`` and
+       ``CHPL_COMM_SUBSTRATE=ofi`` (see also :ref:`readme-omnipath`)
+
```

Comment:
Maybe add

> * If you will not use multi-locale runs at all, set `CHPL_COMM=none`, or compile your application with `--local`. Otherwise, the compiler will generate an executable that can handle multi-locale, which may incur some unnecessary overheads in a single-locale run.

Reply:
Added

 * If you are not using multiple locales, set ``CHPL_COMM=none`` or
   compile your application with `--local` for the best performance.